rfc/terraform-provider.md

```diff
@@ -0,0 +1,92 @@
+# Constellation Terraform Provider
+
+The Constellation Terraform Provider allows its user to manage the full lifecycle of a Constellation cluster -- namely creation, updates, and deletion -- via Terraform.
+
+## Design Goals
+
+- The User needs to be able to perform Creation, Upgrades and Deletion of a Cluster through the provider.
+- Infrastructure provisioning should explicitly **not** be performed by the Constellation Terraform provider.
+- The User needs to receive some sort of Access Token to the Cluster (e.g. a Kubeconfig, whether long- or short-lived)
+when applying a configuration containing the provider and the resources listed below.
+
+## Terraform Configuration Layout for the Constellation Resource
+
+This resembles an examplary configuration of a Constellation cluster through Terraform. While the exact naming and struct layout
+of the individual components may change until the implementation, the structure of the overall configuration looks like this:
+
+```hcl
+terraform {
+  required_providers {
+    constellation = {
+      source  = "tbd/constellation"
+      version = "2.13.0"
+    }
+  }
+}
+
+provider "constellation" { }
+
+resource "constellation_cluster" "foo" {
+    uid = "bar"
+    name = "baz"
+    image = data.constellation_image.ref # or provide manually crafted values
+    kubernetes_version = "v1.27.6"
+    constellation_microservice_version = "lockstep" # or pinned to a specific version
+    debug = false
+    init_endpoint = "10.10.10.10" # should use public ip of LB resource, ideally also provisioned through TF
+    kubernetes_api_endpoint = "10.10.10.10" # should use public ip of LB resource, ideally also provisioned through TF
+    extra_microservices = {
+        csi_driver = true
+        # + more
+        # possiblly also constellation microservices with version and maybe service options,
+        # which would make constellation_microservice_version obsolete.
+        # exact API TBD
+    }
+    master_secret = "foo" # updating this would force recreation of the cluster
+    init_secret = "bar" # maybe derive from master_secret, updating this would force recreation of the cluster
+    network_config = {
+        # TBD
+        # should contain CIDRs for pod network, service cidr, node network... for Cilium
+        # and in-cluster Kubernetes API endpoint, e.g. for Kubelets
+    }
+    attestation = data.constellation_attestation.attestation # or provide manually crafted values
+}
+
+# constellation_cluster provides:
+# constellation_cluster.foo.kubeconfig
```

Comment:
I think the kubeconfig would just be stored in the output after init

Reply:
This would be the case if we tried to support importing resources. I don't know if we need to or if we want to yet, but in that case, we would have to make the Kubeconfig an input or find another way to pass it when importing the resource. For the init-case, we can make it so that the resource is only stored on init, and just returned from the state on any other operation. We cannot trivially recover the Kubeconfig with the mastersecret.

Reply:
Got it: the kubeconfig is persisted in TF state, because it is part of the output, and this is where the provider gets it from for upgrades etc.